lib/recursive-methods.rb

```diff
@@ -1,49 +1,67 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  raise ArgumentError if n < 0
+  return 1 if n == 0
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n), where n is the length of the string
+# Space complexity: O(n), where n is the length of the string
+def reverse(s, pointer = s.length - 1)
+  return s if s == ""
+  return s[pointer] if pointer == 0
+  return s[pointer] + reverse(s, pointer - 1)
```

Comment:
String concatenation creates a new array and  copies all the individual elements over and so is O(n) by itself.